maven embedder workitem

tsurdilo opened the pull request:
Workitem that allows execution of maven commands. Uses Maven Embedder.
Goes along nicely with the git and jira workitems.
See unit test for how to use and what it does.

mswiderski commented on mavenembedder-workitem/src/main/java/org/jbpm/process/workitem/mavenembedder/MavenEmbedderWorkItemHandler.java:
```diff
@@ -0,0 +1,121 @@
+        parameters = {
+                @WidParameter(name = "Commands", required = true),
+                @WidParameter(name = "WorkDirectory", required = true),
+                @WidParameter(name = "ProjectRoot")
```
I'd say it should be required as well

mswiderski commented on mavenembedder-workitem/src/main/java/org/jbpm/process/workitem/mavenembedder/MavenEmbedderWorkItemHandler.java:
```diff
@@ -0,0 +1,121 @@
+            String workDir = (String) workItem.getParameter("WorkDirectory");
+            String projectRoot = (String) workItem.getParameter("ProjectRoot");
+            if(projectRoot == null || projectRoot.length() < 1) {
+                projectRoot = "$M2_HOME";
```
why setting project root to maven home?

tsurdilo replied:
this is where .mvn can be found if needed.

tsurdilo replied:
i will remove this since projectRoot is going to be passed in required field

mswiderski commented on mavenembedder-workitem/src/main/java/org/jbpm/process/workitem/mavenembedder/MavenEmbedderWorkItemHandler.java:
```diff
@@ -0,0 +1,121 @@
+        defaultHandler = "mvel: new org.jbpm.process.workitem.mavenembedder.MavenEmbedderWorkItemHandler()",
+        documentation = "${artifactId}/index.html",
+        parameters = {
+                @WidParameter(name = "Commands", required = true),
```
maybe rename this to goals which is usually used when using maven

mswiderski commented on mavenembedder-workitem/src/main/java/org/jbpm/process/workitem/mavenembedder/MavenEmbedderWorkItemHandler.java:
```diff
@@ -0,0 +1,121 @@
+            MavenCli cli = new MavenCli();
+
+            String origMultiModuleProjectRoot = System.getProperty("maven.multiModuleProjectDirectory");
+            System.setProperty("maven.multiModuleProjectDirectory", projectRoot);
```
what these two lines do?

tsurdilo replied:
since maven 3.3.1 maven.multiModuleProjectDirectory prop was added. It used to find the location of .mvn. Maven embedder needs it set. The intent of these lines is to first get any existing setting that might be in the execution env (orig) and the set it to whats passed to the handler via params, then at the end reset it to what it was originally set to we dont overwrite anything after execution of the workitem.

tsurdilo commented:
@mswiderski updated and added tests - thank you! 
Regarding your concern about how long this could take - yes its a valid concern. I have added CLOptions parameters where users can tweak the execution. 
I ran in a local test the jbpm build from this workitem handler with -DskipTests options and it did not take that long (depends on execution environment of course). 
I do understand where this could also be an executor command but for me this is really useful. 
I use it for an app right now that im working on which uses this workitem to call the archetype generate maven command to generate a project jar from archetype and inputs from business process.

mswiderski commented:
@tsurdilo the main concern is about transaction timeouts when running the build in sync way - take into account download time for dependencies etc. but if we could make it configurable that would be perfect.

tsurdilo commented:
@mswiderski updated to handler to add a sync/async options similar to whats in 
org.jbpm.process.workitem.webservice.WebServiceWorkItemHandler
Hope this is better now please let me know. Thank you!

tsurdilo commented:
@mswiderski added command and test. thanks!!

mswiderski commented on mavenembedder-workitem/src/main/java/org/jbpm/process/workitem/mavenembedder/MavenEmbedderWorkItemHandler.java:
```diff
@@ -0,0 +1,148 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(MavenEmbedderWorkItemHandler.class);
+    private static final String RESULTS_VALUES = "MavenResults";
+    private KieSession kieSession;
```
remove this ksession as it's the old way of doing things

mswiderski commented on mavenembedder-workitem/src/main/java/org/jbpm/process/workitem/mavenembedder/MavenEmbedderWorkItemHandler.java:
```diff
@@ -0,0 +1,148 @@
+    private KieSession kieSession;
+
+    public MavenEmbedderWorkItemHandler(KieSession kieSession) {
+        this.kieSession = kieSession;
```
ksession is not needed

mswiderski commented on mavenembedder-workitem/src/main/java/org/jbpm/process/workitem/mavenembedder/MavenEmbedderWorkItemHandler.java:
```diff
@@ -0,0 +1,148 @@
+
+                                    manager.disposeRuntimeEngine(engine);
+                                } else {
+                                    kieSession.getWorkItemManager().completeWorkItem(workItemId,
```
remove this else as it will not work anyway - better to throw exception if manager is not found - at least from process execution point of view

tsurdilo commented:
@mswiderski updated - thanks again!
